### ISSUE TITLE
How would I create a custom Zsh Completion, but for another shell?

### ISSUE THREAD

**Issue by dfinke** [spf13/cobra#771]
I would like to take my commands and provide the user a way to auto gen completions in a format for another shell. Are there examples on how to do this?

Thank you

**Comment by ryanhartje** [comment]
This is a bit of a non-answer, so apologies for that;

Here's an example of how helm does it, which is another project using Cobra:
https://github.com/helm/helm/blob/master/scripts/completions.bash

If you notice, there's not really an intersection of this bash completion script, and go or cobra. The commands and arg sets look to be hardcoded in. Not only that, but for me, it's a bit hard to follow.

Being curious since you mentioned it, I wondered how it works for zsh since that's my shell of choice, and I stumbled on this very helpful guide:
https://github.com/ecliptik/dotfiles/blob/master/.zprezto/modules/completion/external/zsh-completions-howto.org

I think this guide articulates how to go about writing a completion script, but still falls short of providing a simple, succinct example. I think that this article provides a good starting point, where you can craft a simple completion script, then turn to the in-depth how to for more advance methods/functionality:
https://askql.wordpress.com/2011/01/11/zsh-writing-own-completion/

**Comment by dfinke** [comment]
Thanks @ryanhartje. I'm new to golang, I saw in the the cobra code base the zsh completion code they use.  I don't know enough about pointer receivers yet. My plan was to take that code and modify it to what I need. Coming from an OOP background, it looked to me that I would need to add my code to the base cobra code in order to traverse the `commands` in my code. 

Is that true? Or can I iterate over the `commands` another way?

**Comment by marckhouzam** [comment]
@dfinke Cobra now supports auto-completion for 4 different shells: `bash`, `zsh`, `fish`, and `powershell`.  If you still need to add support for a new shell, it has become much easier because Cobra now provides a hidden command which can be used to obtain completions from different shell scripts, the `__complete` command.  Both `zsh` and `fish` use this command as their sole source of completions.  An example of `bash` using that same technique can be seen in #1146 .

If this addresses your need, can this issue be closed?

**Comment by johnSchnake** [comment]
Closing due to inactivity. Feel free to reopen if you still need help.
